hs_restclient/__init__.py

```diff
@@ -1 +1,666 @@
+                                                 pid=pid)
+        r = self._request('GET', url)
+        if r.status_code != 200:
+            if r.status_code == 403:
+                raise HydroShareNotAuthorized(('GET', url))
+            elif r.status_code == 404:
+                raise HydroShareNotFound((pid,))
+            else:
+                raise HydroShareHTTPException((url, 'GET', r.status_code))
+
+        resource = r.json()
+        assert(resource['resource_id'] == pid)
+
+        return resource
+
+    def getResource(self, pid, destination=None, unzip=False):
+        """ Get a resource in BagIt format
+
+        :param pid: The HydroShare ID of the resource
+        :param destination: String representing the directory to save bag to. Bag will be saved to file named
+            $(PID).zip in destination; existing file of the same name will be overwritten. If None, a stream to the
+            zipped bag will be returned instead.
+        :param unzip: True if the bag should be unzipped when saved to destination. Bag contents to be saved to
+            directory named $(PID) residing in destination. Only applies when destination is not None.
+
+        :raises: HydroShareArgumentException if any arguments are invalid.
+        :raises: HydroShareNotAuthorized if the user is not authorized to access the
+            resource.
+        :raises: HydroShareNotFound if the resource was not found.
+        :raises: HydroShareHTTPException to signal an HTTP error
+
+        :return: None if the bag was saved directly to disk.  Or a generator representing a buffered stream of the
+            bytes comprising the bag returned by the REST end point.
+        """
+        stream = self._getBagStream(pid)
+        if destination:
+            self._getBagAndStoreOnFilesystem(stream, pid, destination, unzip)
+            return None
+        else:
+            return stream
+
+    def _getBagAndStoreOnFilesystem(self, stream, pid, destination, unzip=False):
+        if not os.path.isdir(destination):
+            raise HydroShareArgumentException("{0} is not a directory.".format(destination))
+        if not os.access(destination, os.W_OK):
+            raise HydroShareArgumentException("You do not have write permissions to directory '{0}'.".format(destination))
+
+        filename = "{pid}.zip".format(pid=pid)
+        tempdir = None
+        if unzip:
+            tempdir = tempfile.mkdtemp()
+            filepath = os.path.join(tempdir, filename)
+        else:
+            filepath = os.path.join(destination, filename)
+
+        # Download bag (maybe temporarily)
+        with open(filepath, 'wb') as fd:
+            for chunk in stream:
+                fd.write(chunk)
+
+        if unzip:
+            dirname = os.path.join(destination, pid)
+            zfile = zipfile.ZipFile(filepath)
+            zfile.extractall(dirname)
+            shutil.rmtree(tempdir)
```

Comment:
Add a try/except so that any errors in extracting the archive still allow the temporary files to be cleaned up.